src/NimBLERemoteCharacteristic.cpp

```diff
@@ -402,6 +401,8 @@ std::string NimBLERemoteCharacteristic::readValue() {
 
     do {
         m_semaphoreReadCharEvt.take("readValue");
+        // Clear the value before reading.
+        m_value = "";
```

Comment:
Why do you clear m_value inside of the loop? Once should be enough.

Reply:
Looking at the [onReadCB](https://github.com/h2zero/NimBLE-Arduino/blob/92c0ca4630c519b1a4f010f9ef3971068aec1dc0/src/NimBLERemoteCharacteristic.cpp#L463) function now, that will always return zero. So in my suggestion it will never called.
But also looking at the CB, I think the value should be cleared if, e.g. the CB is not meant for this connection (the first `if` in the CB). Although I am not sure, either the value should be kept to the existing value, or it should be set with a new, read, value. But never cleared.

Reply:
It needs to be cleared before the call to `ble_gattc_read_long()`, when that is called it transmits the read request to the peripheral and when it gets a response the `onReadCB()` is triggered. In that callback we need to check if the connection id matches so we don't update the wrong characteristic, the `return 0` there just tells NimBLE to continue with the operation. After that we need to `m_value +=` to handle long reads that call the callback multiple times so it needs an empty container on the first call. Also if there was an error reading the value then it makes sense for it to be NULL since no data was read.

Reply:
Thnx for the explanation, as I understand it now `ble_gattc_read_long()` will generate one or more CBs, and in each CB the `m_value` is updated by extension (`+=`). Still I think that it is impossible that `m_value` has been changed if the loop is entered a second time, but this is too minor to discuss about!